linera-views/src/simp_store.rs

```diff
@@ -0,0 +1,333 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! A simplified key value store from which we can implement a key-value store
+
+use crate::{
+    batch::{Batch, DeletePrefixExpander, SimplifiedBatch, SimplifiedBatchIter},
+    common::{from_bytes_opt, KeyIterable, KeyValueIterable, KeyValueStore, MIN_VIEW_TAG},
+};
+use async_trait::async_trait;
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use static_assertions as sa;
+use std::fmt::Debug;
+use thiserror::Error;
+
+/// The tag used for the journal stuff.
+const JOURNAL_TAG: u8 = 0;
+sa::const_assert!(JOURNAL_TAG < MIN_VIEW_TAG);
+
+/// Data type indicating that the database is not consistent
+#[derive(Error, Debug)]
+pub enum JournalConsistencyError {
+    /// The journal block could not be retrieved, it could be missing or corrupted
+    #[error("the journal block could not be retrieved, it could be missing or corrupted")]
+    FailureToRetrieveJournalBlock,
+}
+
+#[repr(u8)]
+enum KeyTag {
+    /// Prefix for the storing of the header of the journal.
+    Journal = 1,
+    /// Prefix for the block entry.
+    Entry,
+}
+
+fn get_journaling_key(base_key: &[u8], tag: u8, pos: u32) -> Result<Vec<u8>, bcs::Error> {
+    // We used the value 0 because it does not collide with other key values.
+    // since other tags are starting from 1.
+    let mut key = base_key.to_vec();
+    key.extend([JOURNAL_TAG]);
+    key.extend([tag]);
+    bcs::serialize_into(&mut key, &pos)?;
+    Ok(key)
+}
+
+/// Low-level, asynchronous key-value operations with simplified batch
+#[async_trait]
+pub trait SimplifiedKeyValueStore {
```

Comment:
Yes, but one problem is that the classical `KeyValueStore` has a `clear_journal` function which is absent from the simplified one, and that the simplified one has a `write_batch` that also takes a `base_key` argument.

We could set some dummy parameters and dummy functions if you so prefer, but the code becomes less clear in my opinion.

Reply:
OK I see. In that case, here is another option. We could define the traits:
* `ReadableKeyValueStore`
* `DirectlyWritableKeyValueStore` (the `SimpleBatch`, does not support journaling)
* `WritableKeyValueStore` (the normal `Batch`, may support journaling)
* `KeyValueStore = ReadableKeyValueStore + WritableKeyValueStore`

The combination `ReadableKeyValueStore + DirectlyWritableKeyValueStore` could be used as a trait bound for the struct `JournalingKeyValueStore` without giving it a special name (but if you must, I suppose it could be named `DirectKeyValueStore` or `SimpleKeyValueStore`). I would keep using `store` for the local variable of all stores, if possible. If you must, I suggest using `direct_store` or `simple_store` (shorter/nicer than `simplified_store`).

Reply:
Thanks, will do.

Reply:
How should I handle the fact that we want the error types from `ReadableKeyValueStore` and `WritableKeyValueStore` are matchin in the construction of `KeyValueStore`?